### PR TITLE
Ensure nexus-repository-pypi provide the data-requires-python attribute

### PR DESCRIPTION
**Pull request by renaudll**
This PR ensure that nexus-repository-pypi provide the "data-requires-python" when necessary. 

It fix:
https://issues.sonatype.org/browse/NEXUS-18117
https://issues.sonatype.org/browse/NEXUS-21150

Per PEP-503 specification:

> A repository MAY include a data-requires-python attribute on a file link. This exposes the Requires-Python metadata field, specified in PEP 345, for the corresponding release. Where this is present, installer tools SHOULD ignore the download when installing to a Python version that doesn't satisfy the requirement.

Without this attributes, Nexus would allow pip to install packages incompatible with the current python environment. Since multiples packages are dropping support for python-2 in favor of python-3, this attribute is very important for clients who still manage python-2 environments.

A good example, is the pytest package which dropped support for python-3 in it's 5.0.0 release.

Here's the link from PyPi:
```
<a href="https://files.pythonhosted.org/packages/b7/a9/e64eae45880d383120ef258e23136c74ecd0757ecae84491b578eabaa562/pytest-5.0.0.tar.gz#sha256=95b700cf21ed5b7e91bce7a6b5a573b2e3ef7b3643d00f681d8f9c4672f9fbdf" data-requires-python=">=3.5">pytest-5.0.0.tar.gz</a><br/>
```

Here's the link on a Nexus server with the bug:
```
<a href="../../packages/b7/a9/e64eae45880d383120ef258e23136c74ecd0757ecae84491b578eabaa562/pytest-5.0.0.tar.gz#sha256=95b700cf21ed5b7e91bce7a6b5a573b2e3ef7b3643d00f681d8f9c4672f9fbdf" rel="internal">pytest-5.0.0.tar.gz</a><br/>
```

And here's the link on a Nexus server with the fix:
```
<a href="../../packages/b7/a9/e64eae45880d383120ef258e23136c74ecd0757ecae84491b578eabaa562/pytest-5.0.0.tar.gz#sha256=95b700cf21ed5b7e91bce7a6b5a573b2e3ef7b3643d00f681d8f9c4672f9fbdf" rel="internal" data-requires-python=">=3.5">pytest-5.0.0.tar.gz</a><br/>
```

I hope this help! Let me know if you would like changes.

### CODE REVIEW

File: plugins/nexus-repository-pypi/src/main/java/org/sonatype/nexus/repository/pypi/internal/PyPiIndexUtils.java [hunk 1]
```diff
@@ -52,59 +51,79 @@
...
       }
     }
     return results;
   }
 
+
+  /**
+   * Convert a link to a map ready to be used for "pypi-index.vm".
+   */
+  static ImmutableMap<String, String> indexLinkToMap(PyPiLink link) {
```
Comment (aornatovskyy) [inline plugins/nexus-repository-pypi/src/main/java/org/sonatype/nexus/repository/pypi/internal/PyPiIndexUtils.java]: According to Sonatype code style method params should be final. would u please set  _PyPiLink link_ as final?

File: plugins/nexus-repository-pypi/src/main/java/org/sonatype/nexus/repository/pypi/internal/PyPiIndexUtils.java [hunk 2]
```diff
@@ -52,59 +51,79 @@
...
+            "link", makePackageLinkRelative(link.getLink()),
+            "file", link.getFile(),
+            "data-requires-python", link.getDataPythonRequires()
+    );
+  }
+
+  /**
+   * Convert a link to a map ready to be used for "pypi-root-index.vm".
+   */
+  static ImmutableMap<String, String> rootIndexLinkToMap(PyPiLink link) {
```
Comment (aornatovskyy) [inline plugins/nexus-repository-pypi/src/main/java/org/sonatype/nexus/repository/pypi/internal/PyPiIndexUtils.java]: According to Sonatype code style method params should be final. would u please set PyPiLink link as final?

File: plugins/nexus-repository-pypi/src/main/java/org/sonatype/nexus/repository/pypi/internal/PyPiIndexUtils.java [hunk 3]
```diff
@@ -52,59 +51,79 @@
...
-          results.put(file, path);
-        }
+        String requiresPython = link.attr(PYPI_REQUIRES_PYTHON);
+        results.add(new PyPiLink(file, path, requiresPython));
       }
     }
     return results;
   }
 
+
```
Comment (aornatovskyy) [inline plugins/nexus-repository-pypi/src/main/java/org/sonatype/nexus/repository/pypi/internal/PyPiIndexUtils.java]: looks like 82 line is needless

File: plugins/nexus-repository-pypi/src/main/java/org/sonatype/nexus/repository/pypi/internal/IndexGroupHandler.java [hunk 4]
```diff
@@ -102,27 +104,26 @@ private String mergeResponses(final String name,
...
     checkNotNull(response);
     checkNotNull(results);
     Payload payload = checkNotNull(response.getPayload());
     try (InputStream in = payload.openInputStream()) {
-      Map<String, String> links = PyPiIndexUtils.extractLinksFromIndex(in);
-      for (Entry<String, String> link : links.entrySet()) {
-        String file = link.getKey();
-        String path = link.getValue();
+      List<PyPiLink> links = PyPiIndexUtils.extractLinksFromIndex(in);
+      for (PyPiLink link: links) {
```
Comment (aornatovskyy) [inline plugins/nexus-repository-pypi/src/main/java/org/sonatype/nexus/repository/pypi/internal/IndexGroupHandler.java]: Trivial: broken formating.   **for (PyPiLink link: links)**

**Comment by orf** [comment]
Does this take into account packages published to Nexus, or only packages proxied?

**Comment by renaudll** [comment]
@aornatovskyy Thanks for the review! I've made the requested changes.

@orf This work for "pypi (hosted)", "pypi (proxy)" and "pypi (group)".
I've done my tests in a python-2 virtualenv and targeted the different repositories using `pip install` with the `--index-url` argument. In all cases, pip recognized the "data-requires-python" attribute.

**Comment by ikus060** [comment]
Coming here since I'm facing a problem related to this attribute. Could we assumed this PR will be merged and release soon ? I'm not familiar with the release process of Nexus and I need to take a decision regarding this bug: either I'm waiting a bit to get this release or I'm compiling Nexus my self with this patch.

**Comment by joedragons** [comment]
@ikus060 My suggestion would generally be to look at the JIRA tickets for status.  If they have fix version, are assigned to a team or are in progress, then it's probably "soon".  If not, it's probably "TBD".

**Comment by santdelv** [comment]
@renaudll Thanks for contributing! , before we can integrate this, could I ask you to sign the [Sonatype Contributor License Agreement](https://sonatypecla.herokuapp.com/sign-cla)

**Comment by renaudll** [comment]
@santdelv Done!

**Comment by ProfOak** [comment]
Hey all. I know it's close to a holiday for some people but I'd like to give this some more visibility. Are there any extra things to be taken care of before merging this?

**Comment by mattjohnson** [comment]
@ProfOak You can follow https://issues.sonatype.org/browse/NEXUS-18117 for updates on this issue. It looks like it is targeted for the next release which should be soon. I hope that helps.

**Comment by ProfOak** [comment]
Thanks for the update and the good news. Cheers. :-)

**Comment by jesco39** [comment]
> @ProfOak You can follow https://issues.sonatype.org/browse/NEXUS-18117 for updates on this issue. It looks like it is targeted for the next release which should be soon. I hope that helps.

Ticket is now closed , @mattjohnson any updates?

**Comment by mattjohnson** [comment]
@jesco39 The fix for NEXUS-18117 was included in the 3.20 release. Note that a bug was introduced between 3.19 and 3.20 that causes some issues with duplicate assets (NEXUS-22770). The workaround for that issue for now is to delete the proxy repository and recreate it. You should start seeing the data-requires attributes after that.

**Pull request closed without merging**